### PR TITLE
[IMP] base: document all field types.

### PR DESCRIPTION
**Pull request by Feyensv**
[Task Id - 2223092](https://www.odoo.com/web#id=2223092&action=333&active_id=2818&model=project.task&view_type=form&cids=1&menu_id=4720)

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/fields.py [hunk 1]
```diff
@@ -2535,6 +2571,16 @@ def _update_inverses(self, records, value):
 
 
 class Many2oneReference(Integer):
+    """ Pseudo-relational field (no FK in database).
+
+    The field value is stored as an id :class:`int` in db.
+
+    Contrary to :class:`Reference` fields, the model has to be specified
+    in a :class:`Char` field, whose name has to be set as `model_field` attribute
```
Comment (AntoineVDV) [inline odoo/fields.py:2579]:
```suggestion
    in a :class:`Char` field, whose name has to be set in the `model_field` attribute
```

File: odoo/fields.py [hunk 2]
```diff
@@ -1246,7 +1250,9 @@ def convert_to_export(self, value, record):
 
 
 class Monetary(Field):
-    """ The decimal precision and currency symbol are taken from the attribute
+    """ This field type encapsulates a :class:`Floats <float>` expressed in a given currency.
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply odoo/fields.py:1253]: I might as well go for a :class:`float` then :D

File: odoo/fields.py [hunk 3]
```diff
@@ -2009,6 +2028,13 @@ def write(self, records, value):
 
 
 class Image(Binary):
+    """This field type encapsulates an image, extending :class:`Binary`.
+
+    :param int max_width: maximum image width
+    :param int max_height: maximum image height
+    :param bool verify_resolution: whether the image resolution should be checked before trying to process it.
+        (default: ``True``)
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply odoo/fields.py:2036]: --> odoo/tools/image.py : 
```
        :param verify_resolution: if True, make sure the original image size is not
            excessive before starting to process it. The max allowed resolution is
            defined by `IMAGE_MAX_RESOLUTION`.
```

File: doc/reference/orm.rst [hunk 4]
```diff
@@ -144,22 +144,27 @@ Advanced Fields
...
 
 When assigning a value to a Date/Datetime field, the following options are valid:
 
 * A `date` or `datetime` object.
-* A string in the proper server format *(YYYY-MM-DD)* for Date fields,
-  *(YYYY-MM-DD HH:MM:SS)* for Datetime fields.
+* A string in the proper server format
+
+  * *(YYYY-MM-DD)* for :class:`~odoo.fields.Date` fields,
+  * *(YYYY-MM-DD HH:MM:SS)* for :class:`~odoo.fields.Datetime` fields.
```
Comment (AntoineVDV) [inline doc/reference/orm.rst:159]:
```suggestion
* A string in the proper server format:

  * ``YYYY-MM-DD`` for :class:`~odoo.fields.Date` fields,
  * ``YYYY-MM-DD HH:MM:SS`` for :class:`~odoo.fields.Datetime` fields.
```

File: odoo/fields.py [hunk 5]
```diff
@@ -1246,9 +1250,12 @@ def convert_to_export(self, value, record):
 
 
 class Monetary(Field):
-    """ The decimal precision and currency symbol are taken from the attribute
+    """ Encapsulates a :class:`float` expressed in a given :class:`res_currency<odoo.addons.base.models.res_currency.Currency>`.
```
Comment (AntoineVDV) [inline odoo/fields.py:1253]:
```suggestion
    """ Encapsulates a :class:`float` expressed in a given :class:`res_currency
    <odoo.addons.base.models.res_currency.Currency>`.
```
No auto-generated link ☃️

File: odoo/fields.py [hunk 6]
```diff
@@ -1246,9 +1250,12 @@ def convert_to_export(self, value, record):
...
+
+    The decimal precision and currency symbol are taken from the ``currency_field`` attribute.
 
-    :param str currency_field: name of the field holding the currency
+    :param str currency_field: name of the :class:`Many2one` field
+        holding the :class:`res_currency <odoo.addons.base.models.res_currency.Currency>`
```
Comment (AntoineVDV) [inline odoo/fields.py:1258]: Same

File: odoo/fields.py [hunk 7]
```diff
@@ -2199,6 +2233,11 @@ def convert_to_export(self, value, record):
 
 
 class Reference(Selection):
+    """ Pseudo-relational field (no FK in database).
+
+    The field value is stored as a :class:`string <str>` following the pattern
+    ``"res_model.res_id"`` :class:`str` in database.
```
Comment (AntoineVDV) [inline odoo/fields.py:2239]:
```suggestion
    The field value is stored as a :class:`string <str>` following the pattern
    ``"res_model.res_id"`` in database.
```

File: odoo/fields.py (hunk 5 above)
Comment (Feyensv) [reply odoo/fields.py:1253]: It's normal, res_currency doc isn't auto generated anywhere, but when we do, the link will be generated :stuck_out_tongue:.

Comment (AntoineVDV) [reply odoo/fields.py:1253]: Still, break the line as suggested :p

File: odoo/fields.py [hunk 8]
```diff
@@ -2049,7 +2078,13 @@ def _process_related(self, value):
 
 
 class Selection(Field):
-    """
+    """ Encapsulates an exclusive choice between different values.
+
+    The choices are specified as a list of tuples(`x`, `Y`) where
```
Comment (mart-e) [inline odoo/fields.py:2083]: isn't this is redundant with the `param selection` below?

Comment (Feyensv) [reply odoo/fields.py:2083]: kind of, I found the Selection field worth a little comment, but in the end, it somehow says the same thing so I should just remove it.

**Comment by Feyensv** [comment]
@robodoo r+

**Pull request closed without merging**